File: leetcode2326/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

/*
Solution:

Time: O(k) | Space: O(m * n)

Runtime: 162 ms | 65.87%
Memory: 130.56 MB | 65.45%

- k: the number of nodes in the linked list 'head'
*/
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        ListNode* current = head;
        int r = 0;
        int c = 0;
        int direction = 0;
        int steps[5] = {0, 1, 0, -1, 0};

        vector<vector<int>> res(m, vector<int>(n, -1));

        while (current) {
            res[r][c] = current->val;
            current = current->next;
            int new_r = r + steps[direction];
            int new_c = c + steps[direction + 1];

            if (new_r >= m || new_r < 0 ||
                new_c >= n || new_c < 0 ||
                res[new_r][new_c] != -1) {
                    direction = (direction + 1) % 4;
                }

            r = r + steps[direction];
            c = c + steps[direction + 1];
        }

        return res;
    }
};

File: leetcode2326/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

/*
Solution:

Time: O(k) | Space: O(m * n)

Runtime: 85 ms | 88.00%
Memory: 10.68 MB | 68.00%

- k: number of nodes in linked list 'head'
*/
impl Solution {
    pub fn spiral_matrix(m: i32, n: i32, head: Option<Box<ListNode>>) -> Vec<Vec<i32>> {

        let mut current = head;
        let mut r: i32 = 0;
        let mut c: i32 = 0;

        let mut direction: usize = 0;
        let steps: [i32; 5] = [0, 1, 0, -1, 0];

        let mut res: Vec<Vec<i32>> = vec![vec![-1; n as usize]; m as usize];

        while let Some(node) = current.take() {
            res[r as usize][c as usize] = node.val;
            current = node.next;

            let new_r: i32 = r + steps[direction];
            let new_c: i32 = c + steps[direction + 1];

            if new_r >= m || new_r < 0 ||
                new_c >= n || new_c < 0 ||
                res[new_r as usize][new_c as usize] != -1 {
                    direction = (direction + 1) % 4;
                }

            r = (r + steps[direction] + m) % m;
            c = (c + steps[direction + 1] + n) % n;
        }

        res
    }
}